packages/ember-runtime/lib/system/core_object.js

```diff
@@ -45,198 +42,135 @@ const reopen = Mixin.prototype.reopen;
 
 export const POST_INIT = symbol('POST_INIT');
 
-function makeCtor() {
-  // Note: avoid accessing any properties on the object since it makes the
-  // method a lot faster. This is glue code so we want it to be as fast as
-  // possible.
+const wasApplied = new WeakMap();
 
-  let wasApplied = false;
-  let initProperties, initFactory;
+const factoryMap = new WeakMap();
 
-  class Class {
-    constructor() {
-      if (!wasApplied) {
-        Class.proto(); // prepare prototype...
-      }
-
-      if (arguments.length > 0) {
-        initProperties = [arguments[0]];
-      }
+const NAME_KEY_MAP = new WeakMap();
+const GUID_KEY_MAP = new WeakMap();
 
-      this.__defineNonEnumerable(GUID_KEY_PROPERTY);
-      let m = meta(this);
-      let proto = m.proto;
-      m.proto = this;
+const prototypeMixinMap = new WeakMap();
+const classMixinMap = new WeakMap();
 
-      if (initFactory) {
-        m.factory = initFactory;
-        initFactory = null;
-      }
-      if (initProperties) {
-        // capture locally so we can clear the closed over variable
-        let props = initProperties;
-        initProperties = null;
+/**
+  @class CoreObject
+  @public
+*/
+class CoreObject {
+  static _initFactory(factory) {
+    factoryMap.set(this, factory);
+  }
 
-        let concatenatedProperties = this.concatenatedProperties;
-        let mergedProperties = this.mergedProperties;
-        let hasConcatenatedProps = concatenatedProperties && concatenatedProperties.length > 0;
-        let hasMergedProps = mergedProperties && mergedProperties.length > 0;
+  constructor(properties) {
+    // pluck off factory
+    let initFactory = factoryMap.get(this.constructor);
+    if (initFactory !== undefined) {
+      factoryMap.delete(this.constructor);
+    }
 
-        for (let i = 0; i < props.length; i++) {
-          let properties = props[i];
+    // prepare prototype...
+    this.constructor.proto();
 
-          assert(
-            'EmberObject.create only accepts objects.',
-            typeof properties === 'object' || properties === undefined
-          );
+    this.__defineNonEnumerable(GUID_KEY_PROPERTY);
```

Comment:
I don't think we need to do this _and_ have the getter below on line 511. Defining the GUID_KEY property directly on the instance does work with generateGuid, prevents it from going up the prototype chain, so maybe it's more performant?

Reply:
It shouldn’t be needed in Ember 3+ since we _always_ use WeakMap’s for the guid (we only needed a property on the object in < 3 to support browsers without weakmaps)